### PR TITLE
Check_Number_Type

### LINKED ISSUE #17: Challenge_check_numbers_even_odd
Write a function that prompts a question to enter a number and then it checks and then prints the type of the number, if it's even or odd 
Example:
input: 22
output: "The number is even"

### PR DESCRIPTION
**Pull request by Emanfalouji**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_check_number_type.py
Comment (Frank2446-dotcom) [inline solutions/tests/test_check_number_type.py:1]: The test cases are comprehensive, covering a wide range of scenarios such as whitespace, negative numbers, and invalid inputs. Excellent work ensuring edge cases and defensive tests are addressed!

File: solutions/check_number_type.py
Comment (Frank2446-dotcom) [inline solutions/check_number_type.py:1]: The function is well-structured and clearly handles user input validation, including empty and non-numeric cases.

**Review by Frank2446-dotcom — COMMENTED** [review]
Well done! The implementation and tests are clear and handle the required cases effectively.

**Comment by Emanfalouji** [comment]
Hi 
Could you review the code using the checklist please

File: solutions/check_number_type.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+    Parameters:
+        user_input (str):  str
+    Raises:
+        AssertionError: if the argument is not a integer or empty
+
+    Returns:
+        results will be a text whether "The number is even", "The number is odd"
+        or "Enter a valid number "
+        Examples :
```
Comment (AzzaOmer1) [inline solutions/check_number_type.py:27]: The docstring includes only two  doctests passing , Includes 3 or more passing doctests

File: solutions/check_number_type.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+
+def check_number_type(user_input: str) -> str:
+    """
+    The function asks the user to enter a number and determines if it is type (even or odd.)
+
+
+    Parameters:
+        user_input (str):  str
+    Raises:
```
Comment (AzzaOmer1) [inline solutions/check_number_type.py:20]: Each assertion checks for only one assumption about the argument, each argument have one like,
    ValueError: .....
    ValueError: .....

**Review by AzzaOmer1 — CHANGES_REQUESTED** [review]

**Comment by Emanfalouji** [comment]
Thank you for your insightful commits @AzzaOmer1 
Please take a look at the edits

File: solutions/check_number_type.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+"""
+
+
+def check_number_type(user_input: str) -> str:
+    """
+    The function asks the user to enter a number and determines if it is type (even or odd.)
+
+
+    Parameters:
+        user_input (str):  str
```
Comment (AzzaOmer1) [inline solutions/check_number_type.py]: explain more for input parameter

**Review by AzzaOmer1 — CHANGES_REQUESTED** [review]

File: solutions/check_number_type.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        results will be a text whether "The number is even", "The number is odd"
+        or raises an appropriate error.
+        Examples :
+        >>> check_number_type("20")
+        "The number is even"
+        >>> check_number_type("11")
+        "The number is odd"
+        >>> check_number_type("-11")
+        "The number is odd"
+        >>> check_number_type("")
```
Comment (AzzaOmer1) [inline solutions/check_number_type.py:34]: Detail input assumptions and expand the Raises section to document defensive assertions thoroughly.

File: solutions/check_number_type.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        "Input cannot be empty. Enter a valid number."
+        >>> check_number_type("Eman")
```
Comment (AzzaOmer1) [inline solutions/check_number_type.py:38]: Detail input assumptions and expand the Raises section to document defensive assertions thoroughly.

**Review by AzzaOmer1 — CHANGES_REQUESTED** [review]

**Review by AzzaOmer1 — APPROVED** [review]

**Comment by AzzaOmer1** [comment]
done all reviews
